swap/src/recover.rs

```diff
@@ -134,10 +134,87 @@ pub async fn alice_recover(
                 }
             };
         }
-        Alice::BtcRedeemable { redeem_tx, .. } => {
-            bitcoin_wallet
-                .broadcast_signed_transaction(redeem_tx)
-                .await?;
+        Alice::BtcRedeemable { redeem_tx, state } => {
+            let tx_lock_height = bitcoin_wallet
+                .transaction_block_height(state.tx_lock.txid())
+                .await;
+
+            let block_height = bitcoin_wallet.0.block_height().await?;
+            let refund_absolute_expiry = tx_lock_height + state.refund_timelock;
+
+            // bob cannot cancel
+            if block_height > refund_absolute_expiry {
+                bitcoin_wallet
+                    .broadcast_signed_transaction(redeem_tx)
+                    .await?;
+            } else {
+                let tx_cancel = TxCancel::new(
+                    &state.tx_lock,
+                    state.refund_timelock,
+                    state.a.public(),
+                    state.B.clone(),
+                );
+
+                let tx_cancel_height = bitcoin_wallet
```

Comment:
📣 I think we're missing broadcasting the cancel transaction. If we just wait for the other party to do it, it may never happen. I think we can simply broadcast it unconditionally just before this line, because we know that it can be included on the blockchain i.e. timelock has been reached.